File: app/src/test/java/com/loabletech/bladewatch/daemon/PearMuxTest.kt
package net.bladewatch.app.daemon

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Test

/**
 * BladeWatch-rdtj.6: the wire format the companion (BladeWatch-rdtj.8) must speak byte for byte.
 * The exact bytes are pinned, not just the round trip, so a change here fails loudly instead of
 * silently breaking every paired companion.
 */
class PearMuxTest {

    @Test
    fun `frames have the documented layout`() {
        assertArrayEquals(byteArrayOf(1, 0, 0, 0, 7, 1), PearMux.open(7))
        assertArrayEquals(byteArrayOf(2, 0, 0, 1, 0, 'h'.code.toByte(), 'i'.code.toByte()), PearMux.data(256, "hi".toByteArray()))
        assertArrayEquals(byteArrayOf(3, 0x7f, -1, -1, -1), PearMux.close(Int.MAX_VALUE))
        assertArrayEquals(byteArrayOf(4, 0, 0, 0, 2, 0, 1, 0, 0), PearMux.window(2, 65_536))
    }

    @Test
    fun `every frame round-trips`() {
        val data = PearMux.decode(PearMux.data(5, byteArrayOf(9, 8, 7)))!!
        assertEquals(PearMux.DATA, data.type)
        assertEquals(5, data.stream)
        assertArrayEquals(byteArrayOf(9, 8, 7), data.payload)
        assertEquals(1_234, PearMux.credit(PearMux.decode(PearMux.window(5, 1_234))!!))
        assertEquals(PearMux.VERSION, PearMux.decode(PearMux.open(5))!!.payload[0])
        assertEquals(PearMux.CLOSE, PearMux.decode(PearMux.close(5))!!.type)
    }

    @Test
    fun `malformed frames decode to null instead of throwing`() {
        assertNull(PearMux.decode(byteArrayOf()))
        assertNull(PearMux.decode(byteArrayOf(2, 0, 0, 0, 1))) // DATA with no payload
        assertNull(PearMux.decode(byteArrayOf(4, 0, 0, 0, 1, 0, 0, 0, 0))) // zero credit
        assertNull(PearMux.decode(byteArrayOf(4, 0, 0, 0, 1, -1, -1, -1, -1))) // negative credit
        assertNull(PearMux.decode(byteArrayOf(3, 0, 0, 0, 1, 0))) // CLOSE with a payload
        assertNull(PearMux.decode(byteArrayOf(9, 0, 0, 0, 1))) // unknown type
        assertNull(PearMux.decode(ByteArray(5 + PearMux.MAX_DATA + 1).also { it[0] = PearMux.DATA }))
    }

    @Test
    fun `oversized or empty DATA is refused at encode time`() {
        assertThrows(IllegalArgumentException::class.java) { PearMux.data(1, ByteArray(0)) }
        assertThrows(IllegalArgumentException::class.java) { PearMux.data(1, ByteArray(PearMux.MAX_DATA + 1)) }
        assertThrows(IllegalArgumentException::class.java) { PearMux.window(1, 0) }
    }
}

File: companion/test/transport/pear_mux_test.dart
import 'dart:typed_data';

import 'package:bladewatch_companion/transport/pear_mux.dart';
import 'package:flutter_test/flutter_test.dart';

/// BladeWatch-rdtj.8: the companion speaks the car's mux byte for byte. These vectors are the same
/// ones app/src/test/java/com/loabletech/bladewatch/daemon/PearMuxTest.kt pins on the car side.
void main() {
  test('frames have the documented layout', () {
    expect(PearMux.openFrame(7), [1, 0, 0, 0, 7, 1]);
    expect(PearMux.dataFrame(256, 'hi'.codeUnits), [2, 0, 0, 1, 0, 0x68, 0x69]);
    expect(PearMux.closeFrame(0x7fffffff), [3, 0x7f, 0xff, 0xff, 0xff]);
    expect(PearMux.windowFrame(2, 65536), [4, 0, 0, 0, 2, 0, 1, 0, 0]);
  });

  test('every frame round-trips', () {
    final data = PearMux.decode(PearMux.dataFrame(5, [9, 8, 7]))!;
    expect(data.type, PearMux.data);
    expect(data.stream, 5);
    expect(data.payload, [9, 8, 7]);
    expect(PearMux.decode(PearMux.windowFrame(5, 1234))!.credit, 1234);
    expect(PearMux.decode(PearMux.openFrame(5))!.payload, [PearMux.version]);
    expect(PearMux.decode(PearMux.closeFrame(5))!.type, PearMux.close);
  });

  test('malformed frames decode to null instead of throwing', () {
    for (final bad in <List<int>>[
      [],
      [2, 0, 0, 0, 1], // DATA with no payload
      [4, 0, 0, 0, 1, 0, 0, 0, 0], // zero credit
      [3, 0, 0, 0, 1, 0], // CLOSE with a payload
      [1, 0, 0, 0, 1], // OPEN with no version
      [9, 0, 0, 0, 1], // unknown type
      [2, 0, 0, 0, 1, ...List.filled(PearMux.maxData + 1, 0)],
    ]) {
      expect(PearMux.decode(Uint8List.fromList(bad)), isNull, reason: '$bad'.substring(0, '$bad'.length.clamp(0, 30)));
    }
  });

  test('oversized or empty DATA and non-positive credit are refused at encode time', () {
    expect(() => PearMux.dataFrame(1, const []), throwsArgumentError);
    expect(() => PearMux.dataFrame(1, List.filled(PearMux.maxData + 1, 0)), throwsArgumentError);
    expect(() => PearMux.windowFrame(1, 0), throwsArgumentError);
  });
}
